src/main/kotlin/com/chattriggers/ctjs/api/world/TabList.kt

```diff
@@ -102,17 +123,277 @@ object TabList {
+        val connection = Client.getConnection() ?: return
+        val listedPlayerListEntries = connection.listedPlayerListEntries
+        val playerListEntries = connection.asMixin<ClientPlayNetworkHandlerAccessor>().playerListEntries
+
+        val username = name.unformattedText
+
+        val uuid = UUID.randomUUID()
+        val fakeEntry = PlayerListEntry(GameProfile(uuid, name.unformattedText), false)
+        fakeEntry.displayName = name
+
+        listedPlayerListEntries += fakeEntry
+        playerListEntries[uuid] = fakeEntry
+
+        if (!useExistingSkin) {
+            updateNames()
+            return
+        }
+
+        val mc = Client.getMinecraft()
+        val apiServices =
+            ApiServices.create(mc.asMixin<MinecraftClientAccessor>().authenticationService, mc.runDirectory)
+        apiServices.userCache.setExecutor(mc)
+
+        apiServices.userCache.findByNameAsync(username).thenAcceptAsync {
+            if (it.isPresent) {
+                val result = apiServices.sessionService.fetchProfile(it.get().id, true) ?: return@thenAcceptAsync
+
+                val entry = PlayerListEntry(result.profile, true)
+                entry.displayName = name
+
+                listedPlayerListEntries += entry
+                playerListEntries[result.profile.id] = entry
+
+                listedPlayerListEntries -= fakeEntry
+                playerListEntries.remove(uuid)
+
+                updateNames()
+            }
+        }
+    }
+
+    @JvmStatic
+    @JvmOverloads
+    fun addName(name: String, useExistingSkin: Boolean = true) = addName(TextComponent(name), useExistingSkin)
+
+    /**
+     * Removes all names from the tab list with a certain name
+     *
+     * @param name the name of the entry to remove
+     */
+    @JvmStatic
+    fun removeNames(name: TextComponent) {
+        tabListNames.filter { areTextsEqual(it.getName(), name) }.forEach(Name::remove)
+    }
+
+    private fun updateNames() {
+        tabListNames.clear()
+
+        if (!customHeader)
+            tabListHeader = TextComponent("")
+
+        if (!customFooter)
+            tabListFooter = TextComponent("")
+
+        val hud = toMC()?.asMixin<PlayerListHudAccessor>() ?: return
+        val player = Player.toMC() ?: return
+
+        if (!customHeader)
+            tabListHeader = TextComponent(hud.header)
+
+        if (!customFooter)
+            tabListFooter = TextComponent(hud.footer)
+
+        tabListNames = playerComparator
+            .sortedCopy(player.networkHandler.playerList)
+            .mapTo(mutableListOf(), ::Name)
+    }
+
+    internal fun resetCache() {
+        needsUpdate = true
+    }
+
+    internal fun clearCustom() {
+        tabListNames.clear()
+        customHeader = false
+        customFooter = false
+        tabListHeader = TextComponent("")
+        tabListFooter = TextComponent("")
+    }
+
+    private fun areTextsEqual(a: Text, b: Text): Boolean {
```

Comment:
Maybe `operator fun Text.equals(other: Text)` would be cleaner? Up to you though. Could also just inline this since it's only used once